File: examples/elm-core/ir-spec/result.ml
type ('tError, 'tValue) result = Ok of 'tValue | Err of 'tError

let rec with_default'2 : 'tA * (_, 'tA) result -> 'tA = function
  | v__Def, Ok v_A -> v_A
  | v_Def, Err v__E -> v_Def

let rec with_default'1 : 'tA -> (_, 'tA) result -> 'tA = function
  | v_Def -> ( function v_Res -> with_default'2 (v_Def, v_Res) )

let rec map'2 : ('tA -> 'tB) * ('tX, 'tA) result -> ('tX, 'tB) result = function
  | v_Func, Ok v_A -> Ok (v_Func v_A)
  | v__Func, Err v_E -> Err v_E

let rec map'1 : ('tA -> 'tB) -> ('tX, 'tA) result -> ('tX, 'tB) result =
  function
  | v_Func -> ( function v_Res -> map'2 (v_Func, v_Res) )

let rec map2'3 :
    ('tA * 'tB -> 'tC) * ('tX, 'tA) result * ('tX, 'tB) result ->
    ('tX, 'tC) result = function
  | v__Func, Err v_X, _ -> Err v_X
  | v__Func, Ok _, Err v_X -> Err v_X
  | v_Func, Ok v_A, Ok v_B -> Ok (v_Func (v_A, v_B))

let rec map2'1 :
    ('tA * 'tB -> 'tC) ->
    ('tX, 'tA) result * ('tX, 'tB) result ->
    ('tX, 'tC) result = function
  | v_Func -> ( function v_ResA, v_ResB -> map2'3 (v_Func, v_ResA, v_ResB) )

let rec and_then'2 :
    ('tA -> ('tX, 'tB) result) * ('tX, 'tA) result -> ('tX, 'tB) result =
  function
  | v_Callback, Ok v_Value -> v_Callback v_Value
  | v__Callback, Err v_Msg -> Err v_Msg

let rec and_then'1 :
    ('tA -> ('tX, 'tB) result) -> ('tX, 'tA) result -> ('tX, 'tB) result =
  function
  | v_Callback -> ( function v_Res -> and_then'2 (v_Callback, v_Res) )

let rec map_error'2 : ('tX -> 'tY) * ('tX, 'tA) result -> ('tY, 'tA) result =
  function
  | v__F, Ok v_V -> Ok v_V
  | v_F, Err v_E -> Err (v_F v_E)

let rec map_error'1 : ('tX -> 'tY) -> ('tX, 'tA) result -> ('tY, 'tA) result =
  function
  | v_F -> ( function v_Res -> map_error'2 (v_F, v_Res) )

let rec to_maybe'1 : (_, 'tA) result -> 'tA Maybe.maybe = function
  | Ok v_V -> Maybe.Just v_V
  | Err _ -> Maybe.Nothing

let rec from_maybe'2 : 'tX * 'tA Maybe.maybe -> ('tX, 'tA) result = function
  | v__Err, Maybe.Just v_V -> Ok v_V
  | v_Err, Maybe.Nothing -> Err v_Err

let rec from_maybe'1 : 'tX -> 'tA Maybe.maybe -> ('tX, 'tA) result = function
  | v_Err -> ( function v_Maybe -> from_maybe'2 (v_Err, v_Maybe) )

let rec is_ok'1 : (_, _) result -> bool = function
  | Ok _ -> true
  | Err _ -> false

File: examples/elm_core/ir-spec/result.erl
-file("elm_core/src/result.erlt", 1).

-module(result).

-export_type([result/2]).

-export([with_default/2, with_default/1]).

-export([map/2, map/1, map2/3, map2/1]).

-export([and_then/2,
         and_then/1,
         map_error/2,
         map_error/1]).

-export([to_maybe/1, from_maybe/2, from_maybe/1]).

-export([is_ok/1]).

-import_type({maybe, [{maybe, 1}]}).

-type result(Error, Value) :: {969696,
                               result,
                               result,
                               ok,
                               Value} |
                              {969696, result, result, err, Error}.

-spec with_default(A, result(_, A)) -> A.

with_default(_Def, {969696, result, result, ok, A}) ->
    A;
with_default(Def, {969696, result, result, err, _E}) ->
    Def.

-spec with_default(A) -> fun((result(_, A)) -> A).

with_default(Def) ->
    fun (Res) -> with_default(Def, Res) end.

-spec map(fun((A) -> B), result(X, A)) -> result(X, B).

map(Func, {969696, result, result, ok, A}) ->
    {969696, result, result, ok, Func(A)};
map(_Func, {969696, result, result, err, E}) ->
    {969696, result, result, err, E}.

-spec map(fun((A) -> B)) -> fun((result(X,
                                        A)) -> result(X, B)).

map(Func) -> fun (Res) -> map(Func, Res) end.

-spec map2(fun((A, B) -> C), result(X, A),
           result(X, B)) -> result(X, C).

map2(_Func, {969696, result, result, err, X}, _) ->
    {969696, result, result, err, X};
map2(_Func, {969696, result, result, ok, _},
     {969696, result, result, err, X}) ->
    {969696, result, result, err, X};
map2(Func, {969696, result, result, ok, A},
     {969696, result, result, ok, B}) ->
    {969696, result, result, ok, Func(A, B)}.

-spec map2(fun((A, B) -> C)) -> fun((result(X, A),
                                     result(X, B)) -> result(X, C)).

map2(Func) ->
    fun (ResA, ResB) -> map2(Func, ResA, ResB) end.

-spec and_then(fun((A) -> result(X, B)),
               result(X, A)) -> result(X, B).

and_then(Callback,
         {969696, result, result, ok, Value}) ->
    Callback(Value);
and_then(_Callback,
         {969696, result, result, err, Msg}) ->
    {969696, result, result, err, Msg}.

-spec and_then(fun((A) -> result(X,
                                 B))) -> fun((result(X, A)) -> result(X, B)).

and_then(Callback) ->
    fun (Res) -> and_then(Callback, Res) end.

-spec map_error(fun((X) -> Y),
                result(X, A)) -> result(Y, A).

map_error(_F, {969696, result, result, ok, V}) ->
    {969696, result, result, ok, V};
map_error(F, {969696, result, result, err, E}) ->
    {969696, result, result, err, F(E)}.

-spec map_error(fun((X) -> Y)) -> fun((result(X,
                                              A)) -> result(Y, A)).

map_error(F) -> fun (Res) -> map_error(F, Res) end.

-spec to_maybe(result(_, A)) -> maybe:maybe(A).

to_maybe({969696, result, result, ok, V}) ->
    {969696, maybe, maybe, just, V};
to_maybe({969696, result, result, err, _}) ->
    {969696, maybe, maybe, nothing}.

-spec from_maybe(X, maybe:maybe(A)) -> result(X, A).

from_maybe(_Err, {969696, maybe, maybe, just, V}) ->
    {969696, result, result, ok, V};
from_maybe(Err, {969696, maybe, maybe, nothing}) ->
    {969696, result, result, err, Err}.

-spec from_maybe(X) -> fun((maybe:maybe(A)) -> result(X,
                                                      A)).

from_maybe(Err) ->
    fun (Maybe) -> from_maybe(Err, Maybe) end.

-spec is_ok(result(_, _)) -> boolean().

is_ok({969696, result, result, ok, _}) -> true;
is_ok({969696, result, result, err, _}) -> false.



